### DISCUSSION TITLE
exclude cards with specific tags

### DISCUSSION THREAD

**Discussion by mortii** [mortii/anki-morphs#90] [Ideas]
based on issue #83

**Comment by mortii** [comment]
![image](https://github.com/mortii/anki-morphs/assets/15674619/e5f8beb8-d8ee-4795-96ab-73954d613643)

Suggestions for better column names?

**Comment by mortii** [comment]
Thoughts on changing the background color instead?
![Screenshot from 2023-12-07 17-56-30](https://github.com/mortii/anki-morphs/assets/15674619/d1e60cf6-7444-4dab-bf5c-090bf43160a7)


On light-mode I can barely tell that the font-colors have changed, but maybe that is just my color blindness kicking in.
![Screenshot from 2023-12-07 17-57-17](https://github.com/mortii/anki-morphs/assets/15674619/daa6afee-f666-443d-b55c-cc0bd9537ce2)

Changing the background color is uglier, but definitely more visible.

**Comment by aleksejrs** [comment]
Hierarchical tags could use three-state checkboxes, like this:
![Sample](https://github.com/mortii/anki-morphs/assets/1025861/72c4ddcd-eb77-4280-b7a6-e0144a652cd7) (https://www.codeproject.com/Articles/59514/Simple-Tri-State-TreeView)
I don't know if you can make it with Qt.

If you select a parent tag, the subtags would become automatically selected, but their states would be preserved and their checkboxes would be inactive, so you don't clear your selections by accidentally selecting the parent tag.

**But I would rather have a search query for the tags than a GUI selector.** It is much easier to implement, too.

**Reply by mortii** [reply]
> **But I would rather have a search query for the tags than a GUI selector**. It is much easier to implement, too.

Like there is now? or something else?
![image](https://github.com/mortii/anki-morphs/assets/15674619/b95a0fd0-e098-4c7d-8692-f8a221d3a6a8)

**Reply by HQYang1979** [reply]
Does it default to None? I am not a big fan of bullet, I might just like typing the tags, it's simpler. 
The dialog box's becoming too prominent...

**Reply by Vilhelm-Ian** [reply]
The point of the bulet box is to be able to include and exclude tags not just one. There are othehr ways this can be done. Maybe two lists one next to the other. Where you can add and remove tags one would be for excluding the other one for including. Gimmie a second to draw it out

**Reply by mortii** [reply]
> Does it default to None?

Yes

> The dialog box's becoming too prominent...

It is becoming more complicated than I wanted initially....

> Maybe two lists one next to the other. Where you can add and remove tags one would be for excluding the other one for including. Gimmie a second to draw it out

Please do, the more suggestions the better!

**Reply by Vilhelm-Ian** [reply]
![image](https://github.com/mortii/anki-morphs/assets/81630959/d200d187-26d9-499f-b1b0-34275964ba06)

**Reply by mortii** [reply]
@Vilhelm-Ian very interesting! So you would not see a list of all the tags, but only search for them?

**Reply by Vilhelm-Ian** [reply]
as you search for them they would appear in popup like google does

**Reply by mortii** [reply]
@Vilhelm-Ian that is a good idea. Showing all tags by default is probably overkill.

**Reply by mortii** [reply]
@Vilhelm-Ian how do you feel about the radio buttons in my current version? I think it might be too cluttered...

**Reply by Vilhelm-Ian** [reply]
We can probably do that with your desigen and just add search for tag instead of displaying all tags. I just gave the two list as an idea. I am not opinated on it.

**Reply by aleksejrs** [reply]
@mortii,
> Like there is now? or something else?

No, just a piece of text you enter that the add-on would check cards with, like "`tag:tag1 or tag:tag2 or (tag:tag3 tag:tag4)`".

**Reply by mortii** [reply]
@aleksejrs Right now it just uses regex. I don't really want to recreate the anki search query syntax, that would be very complicated i suspect.

![Screenshot from 2023-12-09 16-45-22](https://github.com/mortii/anki-morphs/assets/15674619/d015fc62-6b0f-4b83-9826-002230558827)

**Reply by aleksejrs** [reply]
I meant that anki-morphs should not show tag names at all. Just use Anki's means to check if a card matches the "include" ~~and "exclude"~~ queries entered by the user. IIRC, there was something like "findCards".

**Reply by mortii** [reply]
@aleksejrs oooh, so basically replace some of the note filter option with a anki search query instead? Very interesting idea, I have to chew on that for a while.

**Reply by aleksejrs** [reply]
Yes, with one query, but maybe in a multi-line editor.

**Comment by mortii** [comment]
I added two buttons, 'view all tags' and 'view selected tags'


![Screenshot from 2023-12-08 13-26-30](https://github.com/mortii/anki-morphs/assets/15674619/3efa10b8-2bea-47a0-956f-1ef8763fdf3a)

![Screenshot from 2023-12-08 13-26-25](https://github.com/mortii/anki-morphs/assets/15674619/5cdadb8e-333b-4a44-9826-66e6fda4581d)

**Comment by mortii** [comment]
This is the version I have currently: https://streamable.com/b3zb6f

When you open the tag selector you by default only see the tags you have previously selected.

As for the 'google-style' suggestion @Vilhelm-Ian, for every key stroke it would have to re-fetch tags which might be a lot of unnecessary work, especially if you have 1000+ tags. Right now it just executes once, when you press enter. This is the same way the anki browser does it when you search.

**Reply by Vilhelm-Ian** [reply]
looks good to me. Like the design

**Comment by mortii** [comment]
I have a completely working version now that i pushed to the[ 'exclude' branch](https://github.com/mortii/anki-morphs/tree/exclude).

@Vilhelm-Ian Could you check it out and double check that it is working fine? meta.json has to be deleted. After that I'll do some final cleaning and push a new release.

**Reply by mortii** [reply]
there is a bug where the color does not change after when you click 'unselect all' and then select checkboxes again.

**Reply by mortii** [reply]
fixed the color bug

**Reply by Vilhelm-Ian** [reply]
@mortii I tried it out. The  design is great. I found only one issue. And that is when I created a new tag. It didn't detect it. So I had to restart anki. Something like when we cached names.txt . To fix that it would be a pain. And not really worth the effort.

Next thing. I get what exclude does. But I don't get what include does. I thought it would only show me cards that have tags that are marked as include but it didn't.

**Reply by mortii** [reply]
@Vilhelm-Ian if you exclude 'a' and include 'b' and 'c', then the card can't have 'a' but it also must have 'b' and 'c'. Does that make sense?

**Reply by Vilhelm-Ian** [reply]
yes it makes sense. 
![image](https://github.com/mortii/anki-morphs/assets/81630959/578adbdf-f3e5-44b9-9dcd-4f0a9d4f9fe1)
I guess it could be that I don't have any i+1 cards with that tag.


but the K and V numbers respond to the filter. Before the filter I had 9k K now I have 1K so obviously it works

**Reply by mortii** [reply]
@Vilhelm-Ian Right now the 'tag' column in the tag selector automatically resizes itself to the width of the tags, do you think that is a good idea, or should it just be a fixed size? I don't have long tags so it didn't bother me, but looking at your picture I realize that it can become a little silly when it gets that wide...

**Reply by Vilhelm-Ian** [reply]
@mortii the reason why I have so long tags is because anki has hierachal tags so some premade decks I have have something like this embriology::pathologic::renal::...
. I have no opion on the matter. When I first open the dialog box I didn't notice anything unusual we can just leave it be.

I guess the only thing that can be change is the word Tag to be centered

**Reply by mortii** [reply]
@Vilhelm-Ian 'Tag' is centered, but the column is so wide it doesn't seem like it is. I'll try setting the width to be a fixed length and see if I like that better.

**Reply by Vilhelm-Ian** [reply]
![image](https://github.com/mortii/anki-morphs/assets/81630959/55e12813-a3d3-4766-9a82-a180325f0747)
I meant for it to be displayed here. I guess you enabled the equivalent of "text-align:center" attribute in qt. so it's most likely a qt issue

**Reply by mortii** [reply]
@Vilhelm-Ian I think the column is so wide it continues behind the 'unselect all' button, that is why it looks off-center. So basically only half of the column is visible.

**Comment by mortii** [comment]
I have been thinking about the suggestion @aleksejrs had about using anki search queries, and I think I have come up with a good compromise that fixes some problems.

It would be absolutely amazing if we could generate a search query output of the note filter, that way we could check which cards it would affect:

![Screenshot from 2023-12-10 12-20-47](https://github.com/mortii/anki-morphs/assets/15674619/d826d833-5f15-4efb-aeeb-b7a6c210d5c8)

![Screenshot from 2023-12-10 12-21-53](https://github.com/mortii/anki-morphs/assets/15674619/ef582758-708c-414c-92cb-c2dd68ab0cbe)

If we could also make work the other way around that would be even more amazing, e.g., I create a search query in the anki browser that gets me the cards I want, then I paste that search query in to a note-filter and it automatically sets the options that fits the search query. That would basically be the best of both worlds, if someone wants to use the gui they can, if someone wants to just type it in they can.

Is this feasible? It would basically require writing a two-way compiler, so probably not, lol. But I wanted to at least document the idea.

**Reply by aleksejrs** [reply]
@mortii, a search query can be complex like `(tag -(tag -tag or notetype) tag -tag) notetype`, so I guess you can only extract the notetype not inside any inner parentheses.

**Reply by mortii** [reply]
@aleksejrs perfect example of why this would be a nightmare to implement :laughing:

**Reply by Vilhelm-Ian** [reply]
I think I could implement it. But currently busy with advent of code. I will try after the 25th. Not intrested in this feature but if it will help the project I will give it a swing

**Reply by mortii** [reply]
@Vilhelm-Ian no no no no, don't work on this if you don't desperately want it. You would have to take into account order of operations and parse-trees and whatnot. 

I created a new discussion for this topic: #93
It can just be left there to simmer for a while, if we find that it is unattainable or too impractical then we can abandon it.
